v2/component/launcher.go

```diff
@@ -0,0 +1,438 @@
+	}
+	if empty(o.MLMDServerPort) {
+		return err("MLMDServerPort")
+	}
+	return nil
+}
+
+// NewLauncher creates a new launcher object using the JSON-encoded runtimeInfo
+// and specified options.
+func NewLauncher(runtimeInfo string, options *LauncherOptions) (*Launcher, error) {
+	if err := options.validate(); err != nil {
+		return nil, err
+	}
+
+	bc, err := parseBucketConfig(options.PipelineRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	rt, err := parseRuntimeInfo(runtimeInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// Placeholder replacements.
+	pr := make(map[string]string)
+
+	metadata, err := metadata.NewClient(options.MLMDServerAddress, options.MLMDServerPort)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Launcher{
+		options:                 options,
+		placeholderReplacements: pr,
+		runtimeInfo:             rt,
+		metadata:                metadata,
+		bucketConfig:            bc,
+	}, nil
+}
+
+func (l *Launcher) prepareInputs(ctx context.Context) error {
+	bucket, err := blob.OpenBucket(context.Background(), l.bucketConfig.bucketURL())
+	if err != nil {
+		return fmt.Errorf("Failed to open bucket %q: %v", l.bucketConfig.bucketName, err)
+	}
+	defer bucket.Close()
+
+	// Read input artifact metadata.
+	for k, v := range l.runtimeInfo.InputArtifacts {
+		if len(v.FileInputPath) == 0 {
+			return fmt.Errorf("Missing input artifact metadata file for input: %q", k)
+		}
+
+		b, err := ioutil.ReadFile(v.FileInputPath)
+		if err != nil {
+			return fmt.Errorf("Failed to read input artifact metadata file for %q: %v", k, err)
+		}
+
+		a := &pb.Artifact{}
+		if err := protojson.Unmarshal(b, a); err != nil {
+			return fmt.Errorf("Failed to unmarshall input artifact metadata for %q: %v", k, err)
+		}
+
+		v.Artifact = a
+
+		// Prepare input uri placeholder.
+		key := fmt.Sprintf(`{{$.inputs.artifacts['%s'].uri}}`, k)
+		l.placeholderReplacements[key] = v.Artifact.GetUri()
+
+		// Prepare input path placeholder.
+		v.LocalArtifactFilePath = path.Join("/tmp/kfp_launcher_inputs", k, "data")
+		key = fmt.Sprintf(`{{$.inputs.artifacts['%s'].path}}`, k)
+		l.placeholderReplacements[key] = v.LocalArtifactFilePath
+
+		// Copy artifact to local storage.
+		// TODO: Selectively copy artifacts for which .path was actually specified
+		// on the command line.
+
+		blobKey, err := l.bucketConfig.keyFromURI(v.Artifact.GetUri())
+		if err != nil {
+			return err
+		}
+		r, err := bucket.NewReader(ctx, blobKey, nil)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+
+		if err := os.MkdirAll(path.Dir(v.LocalArtifactFilePath), 0644); err != nil {
+			return err
+		}
+		w, err := os.Create(v.LocalArtifactFilePath)
+		if err != nil {
+			return err
+		}
+		if _, err := io.Copy(w, r); err != nil {
+			return err
+		}
+	}
+
+	// Prepare input parameter placeholders.
+	for k, v := range l.runtimeInfo.InputParameters {
+		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, k)
+		l.placeholderReplacements[key] = v.ParameterValue
+	}
+
```

Comment:
Do we want to resolve `{{$}}` placeholder here? Or we can leave a TODO.

Reply:
Added a TODO.